py/obj.c

```diff
@@ -192,7 +189,7 @@ bool mp_obj_equal(mp_obj_t o1, mp_obj_t o2) {
     if (mp_obj_is_str(o1)) {
         if (mp_obj_is_str(o2)) {
             // both strings, use special function
-            return mp_obj_str_equal(o1, o2);
+            return mp_obj_str_equal(o1, o2) ? local_true : local_false;
         } else {
             // a string is never equal to anything else
```

Comment:
Note that _'a string is never equal to anything else'_ is incompatible with CPython, since a custom class can decide that an instance is equal to a string. Changing this to something like:
```C
    if (mp_obj_is_str(o1) && mp_obj_is_str(o2)) {
        return mp_obj_str_equal(o1, o2) ? local_true : local_false;
    }
```
would make the code smaller and more compatible, at the expense of some speed for unusual comparisons, although it would also require some fix-ups in `objstr.c` since that code has some assumptions about not being passed strings in comparisons.

Reply:
Yeah.  Might be worth doing this in a separate PR/commit, similar to how None is treated (and maybe also int?).

Reply:
Right. I will leave this as is until this PR is merged and then sort out the fixes to make strings work separately.